Add --payload argument

thomasf opened the pull request:
`--dump` just displays the payload as indented json instead of running
  drone-exec.

`--payload` allows merging in a json value as into the generated payload.

Example:

```sh
drone exec  -event tag --payload '{"job":{"id":999}, "build":{"branch":"SOMETHING"}}'  --dump
```

```json
{
   "config": "clone:\n  path: github.com/drone/drone-cli\n\nbuild:\n  image: golang:1.5\n  environment:\n    - GO15VENDOREXPERIMENT=1\n  commands:\n\n    # compile drone for all architectures\n    - GOOS=linux   GOARCH=amd64 go build -o ./bin/linux_amd64/drone   github.com/drone/drone-cli/drone\n    - GOOS=linux   GOARCH=386   go build -o ./bin/linux_386/drone     github.com/drone/drone-cli/drone\n    - GOOS=linux   GOARCH=arm   go build -o ./bin/linux_arm/drone     github.com/drone/drone-cli/drone\n    - GOOS=darwin  GOARCH=amd64 go build -o ./bin/darwin_amd64/drone  github.com/drone/drone-cli/drone\n    - GOOS=windows GOARCH=386   go build -o ./bin/windows_386/drone   github.com/drone/drone-cli/drone\n    - GOOS=windows GOARCH=amd64 go build -o ./bin/windows_amd64/drone github.com/drone/drone-cli/drone\n\n    # tar binary files prior to upload\n    - mkdir dist\n    - tar -cvzf dist/drone_linux_amd64.tar.gz   --directory=bin/linux_amd64   drone\n    - tar -cvzf dist/drone_linux_386.tar.gz     --directory=bin/linux_386     drone\n    - tar -cvzf dist/drone_linux_arm.tar.gz     --directory=bin/linux_arm     drone\n    - tar -cvzf dist/drone_darwin_amd64.tar.gz  --directory=bin/darwin_amd64  drone\n    - tar -cvzf dist/drone_windows_386.tar.gz   --directory=bin/windows_386   drone\n    - tar -cvzf dist/drone_windows_amd64.tar.gz --directory=bin/windows_amd64 drone\n\n    # generate shas for tar files\n    - sha256sum ./dist/drone_linux_amd64.tar.gz   \u003e ./dist/drone_linux_amd64.sha256\n    - sha256sum ./dist/drone_linux_386.tar.gz     \u003e ./dist/drone_linux_386.sha256\n    - sha256sum ./dist/drone_linux_arm.tar.gz     \u003e ./dist/drone_linux_arm.sha256\n    - sha256sum ./dist/drone_darwin_amd64.tar.gz  \u003e ./dist/drone_darwin_amd64.sha256\n    - sha256sum ./dist/drone_windows_386.tar.gz   \u003e ./dist/drone_windows_386.sha256\n    - sha256sum ./dist/drone_windows_amd64.tar.gz \u003e ./dist/drone_windows_amd64.sha256\n\n\npublish:\n  s3:\n    acl: public-read\n    region: us-east-1\n    bucket: downloads.drone.io\n    access_key: $$AWS_KEY\n    secret_key: $$AWS_SECRET\n    source: dist/\n    target: drone-cli/\n    recursive: true\n    when:\n      branch: master\n",
   "secret": "",
   "repo": {
     "id": 0,
     "owner": "",
     "name": "",
     "full_name": "",
     "avatar_url": "",
     "link_url": "https:///github.com/drone/drone-cli",
     "clone_url": "",
     "default_branch": "",
     "timeout": 0,
     "private": true,
     "trusted": false,
     "allow_pr": false,
     "allow_push": false,
     "allow_deploys": false,
     "allow_tags": false
   },
   "build": {
     "id": 0,
     "number": 0,
     "event": "tag",
     "status": "running",
     "enqueued_at": 0,
     "created_at": 0,
     "started_at": 0,
     "finished_at": 0,
     "deploy_to": "",
     "commit": "e90200cf8dade3e162566f6cc16958cb563e4b10",
     "branch": "SOMETHING",
     "ref": "",
     "refspec": "",
     "remote": "",
     "title": "",
     "message": "Add --payload and --dump arguments",
     "timestamp": 0,
     "author": "Thomas Frössman",
     "author_avatar": "",
     "author_email": "thomasf@jossystem.se",
     "link_url": ""
   },
   "build_last": null,
   "job": {
     "id": 999,
     "number": 0,
     "status": "running",
     "exit_code": 0,
     "enqueued_at": 0,
     "started_at": 0,
     "finished_at": 0,
     "environment": {}
   },
   "netrc": null,
   "keys": null,
   "system": {
     "version": "",
     "link_url": "",
     "plugins": [
       "plugins/*",
       "*/*"
     ],
     "globals": [],
     "privileged_plugins": null
   },
   "workspace": null,
   "vargs": null
 }
```

thomasf commented:
I began with another reflection based implementation with a user interface much like the `-e` argument where `--set job.id=222` can be used to modify the payload struct.

The other solution grew in code size very quickly with all the added struct tag parsing, type checking and struct traversals so I went for this much simpler to implement version which should be sufficient for most situations where external scripts and tools needs to quickly specify how a drone build should be executed locally.

bradrydzewski commented on drone/exec.go:
```diff
@@ -141,6 +149,7 @@ func execCmd(c *cli.Context) error {
 
 	var exits []int
 
+axes:
```
any reasons for the `axes` label here? I don't see a breakpoint or goto anywhere

bradrydzewski replied:
edit: nervermind was looking at the wrong revision. I see where it is being used

tboerger replied:
Seriously with goto behavior?

thomasf replied:
Yes, I think I have seen this being done in the stdlib a number of times (I could be wrong). Anyway it's just the Go early return philosophy, I even thnk this is mentioned inneffective Go. The label is there for making the code easier to read.

tboerger replied:
Mkay

tboerger commented:
LGTM

bradrydzewski commented:
I think the purpose `--dump` may not be intuitive based on name. Typically I would expect something like this to be called `--dry-run` but even in this case not sure the name fits.

I also wonder how the `--dump` flag will be used?

thomasf commented:
The `--dump` command is very useful to quickly understand how to set up `--payload`

I agree that the naming of `--dump` isn't all that great.. 

I can split `--dump` into `--verbose` and `--dry-run` instead, does that seem ok?

thomasf commented:
I've made the change mentioned above.

thomasf commented:
So, how does this pull request look now?

I've used the change is this pull request to implement a way to create releases
which in our case involves secrets which are manually decrypted before the build.

This is the simple public drone.yml (https://github.com/alkasir/alkasir/blob/master/.drone.yml)

.drone.yml:

```yml
clone:
  path: github.com/alkasir/alkasir

cache:
  mount:
    - node_modules

services:
  - postgres:9.1

build:
  image: alkasir/alkasir-build-env:0.0.1
  commands:
    - go run make.go ci
```


Somewhere else I have another repo which has these files:

.drone.yml:

```yml
clone:
  path: github.com/alkasir/
  recursive: true

services:
  - postgres:9.1

build:
  image: alkasir-build-env-nonfree:0.0.1
  commands:
    - ./release ci $$ALKASIR_RELEASE_TAG
  environment:
    - ALKASIR_RELEASE_TAG=$$ALKASIR_RELEASE_TAG
    - ALKASIR_CLIENT_VERSION=$$ALKASIR_RELEASE_TAG
    - ALKASIR_CLIENT_DIFF_URL=https://some-url
    - ALKASIR_CLIENT_CENTRAL_URL=https://some-url

publish:
  github_release:
    api_key: MYKEY
    files: alkasir/dist/*

  additional_internal_publishes....
  
```

release:

```sh
#!/bin/bash

set -e

mkdir -p ~/.ssh/
echo StrictHostKeyChecking no >> ~/.ssh/config

# ci-release is executed inside of drone-exec
ci-release() {
  # check args
   if [ $# -eq 0 ]; then
     echo "git tag argument required"
     exit 1
  fi
  local tag=${1}
  shift

  # remove checked out repository if it exists
  rm -rf alkasir

  # clone the repository
  local repo="https://github.com/alkasir/alkasir"
  git clone \
      -q \
      --depth 1 \
      --single-branch \
      -b ${tag} \
      ${repo}

  # enter the workspace
  cd alkasir

  # Switch npm-shrikwrap to our mirror for controlled repeatable builds:
  local npmjsurl="https://registry.npmjs.org/"
  local mirrorurl="https://some-server/content/groups/npm-all/"
  sed -i s@$npmjsurl@$mirrorurl@g npm-shrinkwrap.json

  # start the build
  go run make.go ci

  # [ create pgp signature for release archive files using the alkasir releases gpg key pair ]...
  # [ create binary upgrade diffs and sign them using the upgrade key pair ]...
}

# release is supposed to be run manually to trigger a build running under drone
release(){
  set -x
  if [ $# -eq 0 ]; then
     echo "git tag argument required"
     exit 1
  fi
  local tag=${1}
  shift
  drone exec \
        --verbose \
        -e ALKASIR_RELEASE_TAG=${tag} \
        --deploy \
        --payload '
{
  "repo": {
    "clone_url" : "git://github.com/alkasir/alkasir",
    "owner": "alkasir",
    "name": "alkasir",
    "full_name": "alkasir/alkasir",
    "trusted": true
  },
  "build": {
    "event": "tag",
    "ref": "refs/tags/'"${tag}"'"
  }
}'

}

case ${1} in
  release)
    shift
    release ${*}
    ;;
  ci)
    shift
    ci-release ${*}
    ;;
  *)
    echo "Nothing to do for ${1}"
    echo "Usage: "
    echo "$0 release git-tag-name"
    exit 1
    ;;
esac

```

To trigger a release build I just execute the command

```
./release release 0.0.1
```

There certainly is some room for improvement in this pattern I've started here but I'm quite satisfied with the results so far.

jackspirou commented:
@bradrydzewski and @thomasf if your going to change `--dump` to `--dry-run` have you considered `--plan`?  Not sure if its what your going for, but [`terraform`](https://www.terraform.io/) uses the `--plan` and `--apply` convention.  

Again not sure if it makes sense for this case, but good artists copy, great artists steal :)

thomasf commented:
@jackspirou ,, @bradrydzewski said on gitter.im that he didnt like the --dry-run either...  I'm not sure where to go from here..

> bradrydzewski Feb 12 18:29
I'm honestly not a huge fan of the --dry-run -- it convolutes the code for something I feel like could be in the documentation (ie use this JSON as a baseline)

> thomasf 05:11
@bradrydzewski I can remove the --dry-run and --verbose from the drone-cli pull request.. I still thinik they provide good features for investigating build problems quickly.. The --dry-run feature adds something like 10 LOC so it's not intrusive at all but I can remove them as long as the --payload can be added, that's the most important one for me.
Also not all users know how to read Go code, I would expect the dryrun+verbose would be helpful for people trying to debug more complicated drone configurations

thomasf commented:
I'm using `--payload`, `--verbose` and `--dry-run` almost every day now when testing and setting up new projects for drone.   `--payload` is of course the most important one but the other ones helps alot too, it's a mighty time saver having a short feedback/modification loop.

Any new ideas or comments on this issue?

thomasf commented:
I would like to yet again draw attention to this issue since it implements both a feature which is essential to my drone usage `--payload` and other features which drastically reduces the time it takes for me to set up drone for new projects...

thomasf commented:
I just removed all new arguments except `--payload` and added payload printing to stdout if `-debug` is used instead of the additional `-verbose` argument.

Is this better?

thomasf commented:
I just noticed that debug is true by default for the exec subcommand which means that with this patch the payload is always being printed unless --debug=false is specified..

It's not obvious from the "drone exec -h" that --debug is enabled by default. I would expect such a feature not to be enabled unless explicitly activated by the user..

bradrydzewski commented:
thanks and really sorry for the delay!!!
